calyx-ir/src/control.rs

```diff
@@ -323,6 +446,50 @@ impl Control {
     }
 }
 
+impl StaticControl {
+    /// Convience constructor for empty.
+    pub fn empty() -> Self {
+        StaticControl::Empty(Empty::default())
+    }
+
+    /// Returns the value of an attribute if present
+    pub fn get_attribute<S>(&self, attr: S) -> Option<u64>
+    where
+        S: Into<Id>,
+    {
+        self.get_attributes().get(attr).cloned()
+    }
+
+    /// Returns the value of an attribute if present
+    pub fn get_latency(&self) -> u64 {
+        match self {
+            StaticControl::Enable(StaticEnable { group, .. }) => {
+                group.borrow().get_latency()
+            }
+            StaticControl::Seq(StaticSeq { latency, .. })
+            | StaticControl::Par(StaticPar { latency, .. })
+            | StaticControl::Repeat(StaticRepeat { latency, .. })
+            | StaticControl::If(StaticIf { latency, .. }) => *latency,
+            &StaticControl::Empty(_) => 0,
+        }
+    }
+}
+
+#[derive(Debug)]
+/// Either holds a reference to a StaticControl, reference to a Control, or None
+/// Helpful when we want to be able get get any specific control statement within a
+/// control program. For example, suppose we assign an id to each enable (static or dynamic)
+/// in the control program. A function that takes in an id and returns the appropriate
+/// enable would have to return a GenericControl.
+/// Has the weird affect that GenericControl::Dynamic(Control::Static(_)) can be
+/// a bit redundant with GenericControl::Static(_) but the latter gives us more precise access
+/// to every enum in the static control, instead of just the big wrapper.
+pub enum GenericControl<'a> {
```

Comment:
This is a cool idea! @EclecticGriffin might have thoughts on something like this and if there are alternative representations that would work better

Reply:
Also probably worth definition `From` methods for this `enum` for `Control` and `StaticControl`

Reply:
Also seconding the `From` impl. Rust analyzer can probably generate that for you as well

Reply:
Worth noting that you can setup a special case in the `From` impl for `Control` to turn `Control::Static` into `GenericControl::Static(_)` rather than `GenericControl::Dynamic(_)` effectively automatically unwrapping that. I would generally recommend something like this because that possible redundancy could become really irritating and eliminating it by construction would let you throw an `unreachable!()` on the pattern matches of it

Reply:
Yeah just got rid fo the `None` variant and implemented `From`.